torch/_dynamo/symbolic_convert.py

```diff
@@ -723,7 +723,8 @@ def step(self):
                         # If we really escape from a block and the current
                         # instruction is not in another block, then there
                         # should be no other nested blocks that we are in.
-                        assert len(self.block_stack) == 1
+                        # assert len(self.block_stack) == 1  # this is no longer true?
+                        # pytest -rs -sv test/functorch/test_eager_transforms.py -k test_grad_deprecated_api_cpu
```

Comment:
cc @williamwen42 is this assertion safe to remove?

Reply:
No - why are we removing this?

Reply:
@williamwen42, In the test case below, the length of `block_stack` is 2.

```
import torch
x = torch.randn(())
y = torch.randn(())

@torch.compile(backend='eager', fullgraph=True)
def wrapper_fn(x, y):
    return torch.func.grad(torch.mul)(x, y)

wrapper_fn(x, y)
```

If I comment the assertion, code works just fine.

Reply:
@guilhermeleobas do you have a test case that does not involve grad?

Reply:
We'll need to look more into why block stack is 2 in this case - I don't expect the assertion to fail.

Reply:
@williamwen42, could you elaborate a little bit on why the block stack here is supposed to be 1? Is it related to a context manager?

Reply:
Yes, the block stack is used to keep track of active context managers/try blocks/etc. There is one block for each nested context manager that we are in. The assert is there because if the current instruction is not in any block, we expect that we were previously only 1 context manager/block deep. A block stack size of 2 at this point means we're jumping out of 2 context managers at once, which is something I don't expect to happen.

Reply:
Ok. Perhaps this explains the failure. There are two context managers in `grad`. See:
https://github.com/pytorch/pytorch/blob/ee4d3cac83345ebfecc60ba1a2c38f1a31df005a/torch/_functorch/eager_transforms.py#L1232-L1235

Reply:
I don't expect it because context managers have cleanup bytecode after running, so we should never jump more than 1 context manager out at a time. Example (in particular, the triple `LOAD_CONST None` is context manager cleanup bytecode):
```python
import dis

def fn():
    with X():
        with Y():
            a = b

dis.dis(fn)
```
Output:
```
  3           0 RESUME                   0

  4           2 LOAD_GLOBAL              1 (NULL + X)
             14 PRECALL                  0
             18 CALL                     0
             28 BEFORE_WITH
             30 POP_TOP

  5          32 LOAD_GLOBAL              3 (NULL + Y)
             44 PRECALL                  0
             48 CALL                     0
             58 BEFORE_WITH
             60 POP_TOP

  6          62 LOAD_GLOBAL              4 (b)
             74 STORE_FAST               0 (a)

  5          76 LOAD_CONST               0 (None)
             78 LOAD_CONST               0 (None)
             80 LOAD_CONST               0 (None)
             82 PRECALL                  2
             86 CALL                     2
             96 POP_TOP
             98 JUMP_FORWARD            11 (to 122)
        >>  100 PUSH_EXC_INFO
            102 WITH_EXCEPT_START
            104 POP_JUMP_FORWARD_IF_TRUE     4 (to 114)
            106 RERAISE                  2
        >>  108 COPY                     3
            110 POP_EXCEPT
            112 RERAISE                  1
        >>  114 POP_TOP
            116 POP_EXCEPT
            118 POP_TOP
            120 POP_TOP

  4     >>  122 LOAD_CONST               0 (None)
            124 LOAD_CONST               0 (None)
            126 LOAD_CONST               0 (None)
            128 PRECALL                  2
            132 CALL                     2
            142 POP_TOP
            144 LOAD_CONST               0 (None)
            146 RETURN_VALUE
        >>  148 PUSH_EXC_INFO
            150 WITH_EXCEPT_START
            152 POP_JUMP_FORWARD_IF_TRUE     4 (to 162)
            154 RERAISE                  2
        >>  156 COPY                     3
            158 POP_EXCEPT
            160 RERAISE                  1
        >>  162 POP_TOP
            164 POP_EXCEPT
            166 POP_TOP
            168 POP_TOP
            170 LOAD_CONST               0 (None)
            172 RETURN_VALUE
ExceptionTable:
  30 to 58 -> 148 [1] lasti
  60 to 74 -> 100 [2] lasti
  76 to 98 -> 148 [1] lasti
  100 to 106 -> 108 [4] lasti
  108 to 112 -> 148 [1] lasti
  114 to 114 -> 108 [4] lasti
  116 to 120 -> 148 [1] lasti
  148 to 154 -> 156 [3] lasti
  162 to 162 -> 156 [3] lasti
```

Reply:
Can you share what the bytecode + exception table looks like at the time of the assert fail?

Reply:
Thank you. I just discovered that if I replace the `aux is not None` condition by `has_aux`, this issue gets fixed.
https://github.com/pytorch/pytorch/blob/5a0a964444d6be3b2762d30f5aec456060fa63c8/torch/_functorch/eager_transforms.py#L1287-L1290

I'll continue debugging this, but I guess this won't be a blocker anymore if CI passes.

Reply:
> Can you share what the bytecode + exception table look like at the time of the assert fail?

Sure! here
https://gist.github.com/guilhermeleobas/8503cb67441b56fdf057ce120d31215e

Reply:
Hmm, unsure why block_stack would be 2 here since the only instruction that would cause a block push is BEFORE_WITH and we don't jump backward to it.

Reply:
Should I open a new issue to track this bug? CI is green with latest change and perhaps one can merge this one.

Reply:
Sure. I'll defer the review back to @zou3519.